### ISSUE TITLE
mem_pool without multi-thread safety may cause unknown bugs

### ISSUE THREAD

**Issue by nashif** [zephyriot/zep-jira14#2265]
**_Reported by Quanwen Du:_**

If we look at the code of memory pool, such as {pool_alloc()}, there is no multiple thread protection,  I understood it may be designed to reduce the impact to the realtime scheduling.  but there may be potential issues if we do not support the multi-thread protection. To use {pool_alloc()} as example, I removed parts of codes in {pool_alloc()}), just to show potential issues: Assume at {Code_Line_2},  {alloc_l = 4} & {free_l = 1}, this is possible.  This means, at {Code_Line_1}, {level_empty(p, 4)} is TRUE;  now {Code_Line_2} allocates {blk} successfully, then runs to {Code_Line_3}, but one interrupt happens and another thread is scheduled and that thread frees one buffer. Then schedule back to {Code_Line_3} of this thread.  now {level_empty(p, 4)} becomes FLASE(this is possible as another thread releases one buffer),  now there are 2 issues:
1). at {Code_Line_4}, one much bigger buffer is returned, but the best expected behavior is to return the available buffer at level 4(the just freed one), right?
2). at {Code_Line_5}, a wrong {block->id.level} is returned(the right level is 1, but assigned as 4 wrongly), right? actually {Code_Line_6} is wrong also as {alloc_l} is used wrongly, but it is similar as errors at {Code_Line_5}.

In short, we shall consider to malloc/free the memory pool in a multi-thread safe way. I am worried about there are other potential issues under multi-thread situation.

static int pool_alloc()
{
	for (i = 0; i < p->n_levels; i++) 
	{
		alloc_l = i;
		if (!level_empty(p, i))  /* Code_Line_1 */
		{
			free_l = i;
		}
	}
        /* Code_Line_2 */
	blk = alloc_block(p, free_l, lsizes[free_l]);    
	for (from_l = free_l;
	     level_empty(p, alloc_l) && /* Code_Line_3 */
         from_l < alloc_l; from_l++) 
	{
		blk = break_block(p, blk, from_l, lsizes);
	}
	block->data = blk; /* Code_Line_4 */
	block->id.level = alloc_l;  /* Code_Line_5 */
	block->id.block = ...  /* Code_Line_6 */
	return 0;
}

(Imported from Jira ZEP-2434)

**Comment by nashif** [comment]
**_by Andy Ross:_**

The pool was actually designed to be threadsafe.  I'm having trouble understanding your explanation, honestly.  Do you have a test case that can actually exercise the race you think you see (even if you do it with k_yield() to force a context switch)?

The basic idea of pool_alloc() is that it makes a decision in unsynchronized code about the state of the heap, then attempts to take an action to mutate the heap by calling alloc_block(), which is synchronized with irq_lock().  This step can fail if the heap gets modified in the interrim, and this result is checked and retried.

Honestly I think this is OK from my reading of the code.

**Comment by nashif** [comment]
**_by Quanwen Du:_**

You know, for the race condition, it is not easy to have a reproducible step to make issue happen, so I am not able to show the test case. Actually it is my code review to dig out the issue. Let me embed the comments into the code below, hope you will get what I was saying. Sorry I am not able to upload a picture(my company's policy), otherwise it will be much easier to get understood.  If the code/comment below does not look good, pls copy it out to Notepad++ to get my comments.
I think the safer way is to call irq_lock() to protect the whole logic of pool_alloc, not partially if we do not care about the impact to the real time scheduling, also need to consider other similar codes. There are other ways to protect such logic such as a mutex, k_sched_lock() and etc.

static int pool_alloc()
{
    /* Assume this func is called at ThreadA */
    /* Assume alloc_l = 4 & free_l = 1 due to level_empty(p, 4) = TRUE , this means the logic need to break the pool from level 1 to level 4, then return a best-fit buffer */
    blk = alloc_block(p, free_l, lsizes[free_l]); 
    /* blk is allocated at level[free_l = 1] successfully, yes it is interrupt locked */
    
    /* IMPORTANT: Interrupt is triggered here, and after interrupt, CPU goes to ThreadB */
    /* ThreadB frees a pool buffer, which make level_empty(p, 4) = FALSE */
    /* Now ThreadB is suspended and CPU goes to ThreadA again and come here */
    /* But the for_loop below will NOT be executed at all, due to level_empty(p, 4) = FALSE */
    for (from_l = free_l; level_empty(p, alloc_l) &&  from_l < alloc_l; from_l++)
    { 
        blk = break_block(p, blk, from_l, lsizes); 
    }
    block->data = blk; 
    /* Issue 1:  blk is from level[free_l = 1], NOT from [alloc_l = 4] at all: too big a buffer is returned, because the expected level is 4, but here is 1 */
    block->id.level = alloc_l; 
    /* Issue  2:  the level shall be from_l: blk does not match with its level!*/
    block->id.block = ...  /*Similar as Issue 2, shall use from_l, not alloc_l */
    /* There is a simple fix for Issue 1&2: just to remove [level_empty(p, alloc_l) &&] condition checking from the for_loop above, but shall check other codes such as free_block */

    return 0;
}

**Comment by nashif** [comment]
**_by Andy Ross:_**

Oooh, good catch.  You're right, that level_empty() predicate in the final loop looks wrong for sure.  Once we have the returned big block, we want to break it and use it because we can't rely on the pool state to be consistent.  Will fix.  (Honestly I'm not sure why it's there: I'm guessing this was an attempt at "optimization" on my part to avoid needless block breakage.  I may want to augment break_block() to threadsafely detect the case where the break isn't required.)

But the suggestion to use bigger critical sections and/or blocking primitives can't work.  Remember the original mem_pool was *unsynchronized* because of the latency costs of the big irq_locks, and one of the requirements of the new allocator was that it would be usable from an ISR, which means blocking locks can't be used (we block on a semaphore in OOM conditions, but only when called without a NO_WAIT timeout)

**Comment by nashif** [comment]
**_by Quanwen Du:_**

Yes, I understood the concern about the latency cost if locking the whole malloc procedure, I shared this guess when I submitted the issue, you confirmed this now. And you also shared the ISR scenarios, Thanks. But I am still worried about the multi-thread safety because we need to consider thread scheduling and random ISR very carefully. I think it is a big challenge, anyway I will revisit the topic after you come out a fix.
